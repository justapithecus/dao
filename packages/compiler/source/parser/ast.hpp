#pragma once

#include <memory>
#include <string>
#include <variant>
#include <vector>

namespace dao {

  struct identifier_expr;
  struct numeral_expr;
  struct binary_expr;
  struct function_arg;
  struct function_proto;
  struct function_def;
  struct function_call;

  using ast      = std::variant<numeral_expr, identifier_expr, binary_expr,
    function_proto, function_def, function_call>;
  using ast_node = std::unique_ptr<ast>;

  struct identifier_expr {
    std::string name;
  };

  struct numeral_expr {
    int val;
  };

  struct binary_expr {
    ast_node lhs, rhs;
    char     op;
  };

  struct function_arg {
    std::string name;
  };

  struct function_proto {
    std::string               id;
    std::vector<function_arg> args;
  };

  struct function_def {
    ast_node       body;
    function_proto proto;
  };

  struct function_call {
    std::string           callee;
    std::vector<ast_node> args;
  };

  static_assert(sizeof(ast_node) <= 24);

} // namespace dao
