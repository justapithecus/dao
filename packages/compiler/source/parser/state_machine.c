#include "state_machine.h"

#define offset(kind) ((kind) * (lexical_state_count))

unsigned char const ch_to_eqc[256] = {
  // Layout
  [0]          = offset(glyph_eof),
  [1 ... 9]    = offset(glyph_layout),
  ['\n']       = offset(glyph_new_line),
  ['\r']       = offset(glyph_new_line),
  [11 ... 12]  = offset(glyph_layout),
  [14 ... 31]  = offset(glyph_layout),
  [' ']        = offset(glyph_white_space),
  // Special
  ['!']        = offset(glyph_operator),
  ['(']        = offset(glyph_operator),
  [')']        = offset(glyph_operator),
  ['*']        = offset(glyph_operator),
  ['+']        = offset(glyph_operator),
  ['-']        = offset(glyph_operator),
  ['.']        = offset(glyph_operator),
  ['/']        = offset(glyph_operator),
  ['<']        = offset(glyph_operator),
  ['=']        = offset(glyph_operator),
  ['>']        = offset(glyph_operator),
  // Numbers
  [48 ... 57]  = offset(glyph_number),
  // Letters
  [65 ... 90]  = offset(glyph_letter),
  ['_']        = offset(glyph_letter),
  [97 ... 122] = offset(glyph_letter),
};
