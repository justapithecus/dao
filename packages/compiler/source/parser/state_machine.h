#pragma once

// Glyphs are categories comprising individual characters or strings of characters, e.g.
// '+' and '->', respectively.
typedef enum glyph {
  glyph_layout,
  glyph_white_space,
  glyph_new_line,
  glyph_letter,
  glyph_number,
  glyph_separator,
  glyph_operator,
  glyph_eof,
  glyph_count,
} glyph;

typedef enum lexical_state {
  lexical_state_final,
  lexical_state_next_char,
  lexical_state_separator,
  lexical_state_identifier,
  lexical_state_identifier_end,
  lexical_state_numeral,
  lexical_state_numeral_end,
  lexical_state_operator,
  lexical_state_new_line,
  lexical_state_eof,
  lexical_state_error,
  lexical_state_count,
} lexical_state;

// Character to character-equivalence classes lookup table
//
// Character-equivalence classes are pre-computed offsets used to index into
// the lexical state transitions lookup table.
extern unsigned char const equivalence_class[256];

// Lexical state transitions lookup table
//
// Index using the current state offset by a a character-equivalence class, to lookup
// the next lexical state.
#define LEX_TRANS_SIZE ((glyph_count) * (lexical_state_count))
#pragma clang diagnostic   push
#pragma clang diagnostic   ignored "-Wdeprecated-enum-enum-conversion"
extern unsigned char const transition[LEX_TRANS_SIZE];
#pragma clang diagnostic   pop

// Character rewind
//
// Certain lexical states, typically those terminating a lexeme, requires shifting back 1
// character, to start lexing anew.
extern char const ch_rewind[lexical_state_count];

// True if inside an identifier, value, or operator
extern unsigned char const inside[lexical_state_count];

extern unsigned char const binary_op_precedence[256];
