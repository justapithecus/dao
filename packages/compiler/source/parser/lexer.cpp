#include "../utils.hpp"

#include "lexer.hpp"
#include "state_machine.h"

namespace dao {

  auto lex(std::string_view fpath) -> std::vector<dao::token> {
    std::vector<dao::token> tokens{};
    dao::lexeme             lexeme{};
    auto                    source{dao::fread(fpath.data())};
    auto                    src_ptr{source.data()};
    unsigned char           state{lexical_state_next_char};

    do {
      unsigned char ch;
      auto          eqc = equivalence_class[(ch = *src_ptr++)];

      state = transition[state + eqc];
      src_ptr += ch_rewind[state];
      // lexeme.col_num += ch_rewind[state];
      lexeme.len += inside[state];

      switch (state) {
      case lexical_state_next_char: {
        ++lexeme.col_num;
        break;
      }
      case lexical_state_identifier: {
        lexeme.token_kind = token_kind_identifier;
        break;
      }
      case lexical_state_numeral: {
        lexeme.token_kind = token_kind_numeral;
        break;
      }
      case lexical_state_operator:
        lexeme.token_kind = token_kind_operator;
        [[fallthrough]];
      case lexical_state_identifier_end:
        [[fallthrough]];
      case lexical_state_numeral_end: {
        lexeme.update_repr(src_ptr, 0);
        lexeme.reset();
        tokens.emplace_back(lexeme.as_token());
        break;
      }
      default:
        break;
      }

    } while (state > lexical_state_final);

    return tokens;
  }
} // namespace dao
