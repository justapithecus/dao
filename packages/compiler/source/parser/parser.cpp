#include "parser.hpp"

#include "state_machine.h"

namespace dao {

  parse_context::parse_context(std::vector<token> const &tokens)
    : tokens_{tokens}
    , cursor_{tokens_.begin()} {
  }

  auto parse_context::is_eof() const {
    return cursor_ != tokens_.end() and
           cursor_->kind == token_kind::e_end_of_file;
  }

  auto parse_context::peek() const {
    return cursor_;
  }

  auto parse_context::seek() {
    cursor_ = std::next(cursor_);
    return cursor_;
  }

  auto parse_context::eat() {
    auto tok{cursor_};
    seek();
    return tok;
  }

  auto parse_context::rewind() {
    cursor_ = std::prev(cursor_);
    return cursor_;
  }

  auto parse_context::skip() {
    while (not is_eof() and cursor_->kind == token_kind::e_new_line) {
      seek();
    }
  }

  parser::parser(std::vector<token> const &tokens)
    : ctx_{tokens} {
  }

  auto parser::parse() -> ast {
    program_ast prog{};

    while (not ctx_.is_eof()) {
      switch (ctx_.peek()->kind) {
      case token_kind::e_new_line:
        ctx_.skip();
        break;
      case token_kind::e_keyword_external:
        prog.nodes.emplace_back(parse_external_linkage());
        break;
      case token_kind::e_keyword_function:
        // TODO(andrew): implement control flow, keep existing erroneous behavior for now
        if (auto node{parse_function_def()}) {
          prog.nodes.emplace_back(std::move(node));
          break;
        } else {
          // TODO(andrew): add to ctx.errors
          return prog;
        }

      default:
        if (auto node{parse_expr()}) {
          prog.entry.emplace_back(std::move(node));
          break;
        } else {
          // TODO(andrew): add to ctx.errors
          return prog;
        }
      }
    }

    return prog;
  }

  template <typename T>
  auto parser::parse_literal() -> ast_node {
    auto val{ctx_.eat()->repr};
    return std::make_unique<ast>(T{val});
  }

  template auto parser::parse_literal<numeral_expr>() -> ast_node;
  template auto parser::parse_literal<string_literal>() -> ast_node;

  auto parser::parse_expr() -> ast_node {
    ctx_.skip();

    ast_node node{};

    while (not ctx_.is_eof()) {

      switch (ctx_.peek()->kind) {
      case token_kind::e_new_line:
        // new-line tokens act as terminators
        ctx_.skip();
        [[fallthrough]];
      case token_kind::e_end_of_file:
        if (node) {
          return node;
        }
        break;
      case token_kind::e_keyword_external:
        // TODO(andrew): ctx.errors, currently do not support parsing of external linkages
        //               and functions inside of function definitions
        [[fallthrough]];
      case token_kind::e_keyword_function:
        return nullptr;
      case token_kind::e_keyword_if:
        return parse_if_expr();
      case token_kind::e_keyword_then:
      case token_kind::e_keyword_else:
        // TODO(andrew): may want to have semicolons...
        return node;
      case token_kind::e_identifier:
        node = parse_identifier_expr();
        break;
      case token_kind::e_numeral:
        node = parse_literal<numeral_expr>();
        break;
      case token_kind::e_literal:
        node = parse_literal<string_literal>();
        break;
      case token_kind::e_separator:
        if (ctx_.peek()->repr == "(") {
          node = parse_parenthetical_expr();
        } else if ((ctx_.peek()->repr == ")" or ctx_.peek()->repr == ",") and
                   node) {
          return node;
        } else {
          throw std::runtime_error{"found unexpected separator: " +
                                   ctx_.peek()->repr + " or node was nullptr"};
        }
        break;
      case token_kind::e_operator:
        return parse_binary_expr(std::move(node));
      default:
        // TODO(andrew): add to ctx.errors, encountered unknown
        break;
      }

      if (!node) {
        // TODO(andrew): add to ctx.errors
        return nullptr;
      }
    }

    return node;
  }

  auto parser::parse_primary_expr() -> ast_node {
    switch (ctx_.peek()->kind) {
    case token_kind::e_identifier:
      return parse_identifier_expr();
    case token_kind::e_numeral:
      return parse_literal<numeral_expr>();
    case token_kind::e_literal:
      return parse_literal<string_literal>();
    case token_kind::e_separator:
      if (ctx_.peek()->repr == "(") {
        return parse_parenthetical_expr();
      } else {
        throw std::runtime_error{
          "found unexpected separator: " + ctx_.peek()->repr};
      }
      break;
    case token_kind::e_operator:
      return parse_binary_expr();
    default:
      // TODO(andrew): add to ctx.errors
      return nullptr;
    }
  }

  auto parser::parse_identifier_expr() -> ast_node {
    auto name{ctx_.eat()->repr};

    if (not ctx_.is_eof()) {
      if (ctx_.peek()->as_operand() == '(') {
        return parse_function_call(std::move(name));
      } else if (ctx_.peek()->kind == token_kind::e_operator) {
        auto node{std::make_unique<ast>(identifier_expr{std::move(name)})};
        return parse_binary_expr(std::move(node));
      }
    }

    return std::make_unique<ast>(identifier_expr{std::move(name)});
  }

  auto parser::parse_binary_expr(std::uint8_t op_precedence) -> ast_node {
    auto lhs{parse_primary_expr()};
    if (!lhs) {
      // TODO(andrew): add to ctx.errors
      return nullptr;
    }
    return parse_binary_expr(std::move(lhs), op_precedence);
  }

  auto parser::parse_binary_expr(ast_node lhs, std::uint8_t op_precedence)
    -> ast_node {

    while (not ctx_.is_eof() and ctx_.peek()->kind == token_kind::e_operator) {
      auto op{ctx_.peek()->as_operand()};
      auto token_precedence{
        binary_op_precedence[static_cast<std::uint8_t>(op)]};
      if (token_precedence < op_precedence) {
        return lhs;
      }

      // eat operand
      ctx_.seek();

      auto rhs{parse_primary_expr()};
      if (!rhs) {
        // TODO(andrew): add to ctx.errors
        return nullptr;
      }

      if (ctx_.peek()->kind == token_kind::e_operator) {
        // eat operand
        auto next_op{ctx_.eat()->as_operand()};

        if (not ctx_.is_eof()) {
          auto next_precedence{
            binary_op_precedence[static_cast<std::uint8_t>(next_op)]};
          if (token_precedence <= next_precedence) {
            ctx_.rewind();
            // current right-hand-side becomes the left-hand-side of the inner expression
            rhs = parse_binary_expr(std::move(rhs), token_precedence + 1);
            if (!rhs) {
              // TODO(andrew): add to ctx.errors
              return nullptr;
            }

            if (token_precedence != next_precedence) {
              std::swap(lhs, rhs);
            }
          }
        }
      }

      lhs =
        std::make_unique<ast>(binary_expr{std::move(lhs), std::move(rhs), op});
    }

    return lhs;
  }

  auto parser::parse_parenthetical_expr() -> ast_node {
    // TODO(andrew): expected token error-checking
    // eat '('
    ctx_.seek();
    auto node{parse_expr()};
    // eat ')'
    ctx_.seek();
    return node;
  }

  auto parser::parse_function_arg() -> function_arg {
    auto name{ctx_.eat()->repr};
    return function_arg{std::move(name)};
  }

  auto parser::parse_function_arg_seq() -> std::vector<function_arg> {
    std::vector<function_arg> args{};

    // eat '('
    ctx_.seek();

    // no-args
    if (auto sep{ctx_.peek()->as_operand()}; sep == ')') {
      ctx_.seek();
      return args;
    }

    // TODO(andrew): maybe some kind of generic parse_sequence with separators
    while (not ctx_.is_eof()) {
      args.emplace_back(parse_function_arg());

      // eat ',' or ')'
      if (auto sep{ctx_.peek()->as_operand()}; sep == ')') {
        ctx_.seek();
        return args;
      } else if (sep == ',') {
        ctx_.seek();
      } else {
        // TODO(andrew): unexpected end of function arg sequence
        return args;
      }
    }
    return args;
  }

  auto parser::parse_function_proto() -> function_proto {
    // eat 'function'
    ctx_.seek();

    // eat function identifier
    auto id{ctx_.eat()->repr};
    return function_proto{std::move(id), parse_function_arg_seq()};
  }

  auto parser::parse_function_def() -> ast_node {
    auto proto{parse_function_proto()};

    ctx_.skip();

    // currently do not support defining functions or declaring prototypes
    // inside of function bodies
    if (ctx_.peek()->kind == token_kind::e_keyword_function or
        ctx_.peek()->kind == token_kind::e_keyword_external) {
      return std::make_unique<ast>(std::move(proto));
    }

    auto body{parse_expr()};
    if (not body) {
      return std::make_unique<ast>(std::move(proto));
    }

    return std::make_unique<ast>(
      function_def{std::move(body), std::move(proto)});
  }

  auto parser::parse_expr_seq() -> std::vector<ast_node> {
    std::vector<ast_node> args{};

    // eat '('
    ctx_.seek();

    // no-args
    if (auto sep{ctx_.peek()->as_operand()}; sep == ')') {
      ctx_.seek();
      return args;
    }

    // TODO(andrew): maybe some kind of generic parse_sequence with separators
    while (not ctx_.is_eof()) {
      args.emplace_back(parse_primary_expr());

      // eat ',' or ')'
      if (auto sep{ctx_.peek()->as_operand()}; sep == ')') {
        ctx_.seek();
        return args;
      } else if (sep == ',') {
        ctx_.seek();
      } else {
        // TODO(andrew): unexpected end of function call sequence
        return args;
      }
    }

    return args;
  };

  auto parser::parse_function_call(std::string callee) -> ast_node {
    return std::make_unique<ast>(
      function_call{std::move(callee), parse_expr_seq()});
  }

  ankerl::unordered_dense::map<std::string, linkage_kind> const
    supported_linkages{
      {"C", linkage_kind::e_c_linkage},
    };

  auto parser::parse_external_linkage() -> ast_node {
    // eat 'external'
    ctx_.seek();
    // eat '('
    ctx_.seek();

    // eat <external_linkage_kind>, e.g.: "C"
    auto it{supported_linkages.find(ctx_.eat()->repr)};
    if (it == supported_linkages.end()) {
      // TODO(andrew): ctx.errors, unsupported linkage kind
      return nullptr;
    }
    // eat ')'
    ctx_.seek();

    switch (ctx_.peek()->kind) {
    case token_kind::e_keyword_function:
      return std::make_unique<ast>(
        external_linkage_ast{it->second, parse_function_proto()});
    default:
      // TODO(andrew): ctx.errors, external linkage only supports function declarations
      //               eventually also support symbols to other values
      return nullptr;
    }
  }

  auto parser::parse_if_expr() -> ast_node {
    // eat 'if'
    ctx_.seek();

    auto condition{parse_expr()};
    if (!condition) {
      // TODO(andrew): ctx.errors
      return nullptr;
    }

    if (ctx_.peek()->kind != token_kind::e_keyword_then) {
      // TODO(andrew): expected 'then'
      return nullptr;
    }

    // eat 'then'
    ctx_.seek();
    auto then{parse_expr()};
    if (!then) {
      // TODO(andrew): ctx.errors
      return nullptr;
    }

    ast_node else_{};
    if (ctx_.peek()->kind == token_kind::e_keyword_else) {
      // eat 'else'
      ctx_.seek();
      if (else_ = parse_expr(); not else_) {
        // TODO(andrew): ctx.errors
        return nullptr;
      }
    }

    return std::make_unique<ast>(
      if_expr{std::move(condition), std::move(then), std::move(else_)});
  }

} // namespace dao
