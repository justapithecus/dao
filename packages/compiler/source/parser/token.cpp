#include <iomanip>

#include "state_machine.h"
#include "token.hpp"

#define token_desc(kind, desc)                                                 \
  case (kind):                                                                 \
    os << (desc);                                                              \
    break

namespace dao {

  auto operator<<(std::ostream &os, token const &tok) -> std::ostream & {
    os << std::setw(12);

    switch (tok.kind) {
      token_desc(token_kind_identifier, "Identifier");
      token_desc(token_kind_numeral, "Numeral");
      token_desc(token_kind_operator, "Operator");
      token_desc(token_kind_separator, "Separator");
    default:
      os << "Unknown";
      break;
    }

    return os << " | " << tok.repr << " |";
  }
} // namespace dao
