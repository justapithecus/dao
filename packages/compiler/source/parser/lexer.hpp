#pragma once

#include <iostream>
#include <string_view>
#include <vector>

#include "token.hpp"

namespace dao {

  struct lexeme {
    std::string  repr;
    std::uint8_t len;
    std::uint8_t line_num;
    std::uint8_t col_num;

    explicit lexeme()
      : len{0}
      , line_num{1}
      , col_num{1} {
    }

    auto update_repr(char const *src_ptr, int end_offset) {
      auto first{src_ptr - len};
      auto last{first + (len + end_offset)};
      repr = std::string{first, last};
    }

    // reset trackers for next lexeme
    auto reset(int offset = 0) {
      col_num += (len + offset);
      len = 0;
    }

    auto as_token(dao::token_kind kind) {
      reset();
      return dao::token{repr, kind};
    }
  };

  static_assert(sizeof(dao::lexeme) <= 32);

  auto lex(std::string_view fpath) -> std::vector<token>;
} // namespace dao
