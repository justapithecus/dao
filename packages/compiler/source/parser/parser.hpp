#pragma once

#include <iostream>
#include <iterator>
#include <memory>
#include <type_traits>
#include <vector>

#include "../ast/index.hpp"
#include "token.hpp"

namespace dao {

  using token_cursor = std::vector<token>::const_iterator;

  class parse_context {
    std::vector<token> const &tokens_;
    token_cursor              cursor_;

    // TODO(andrew): errors

  public:
    explicit parse_context(std::vector<token> const &tokens);

    [[nodiscard]]
    auto is_eof() const;

    [[nodiscard]]
    auto peek() const;

    /// seek
    ///
    /// Seeks and returns a cursor to the newly advanced position
    auto seek();

    /// eat
    ///
    /// Eats and returns the value of that token
    [[nodiscard]]
    auto eat();

    /// rewind
    ///
    /// Rewinds and returns a cursor to its previous position
    auto rewind();

    /// skip
    ///
    /// Skips consecutive ineffectual new-line tokens
    auto skip();
  };

  class parser {
    parse_context ctx_;

  public:
    explicit parser(std::vector<token> const &tokens);

    /// Parses tokens into an abstract syntax tree
    ///
    /// <program_ast> ::= { <expr> }
    auto parse() -> ast;

  private:
    /// Parses a literal
    ///
    /// <digit> ::= [0-9]
    /// <numeral_expr> ::= { <digit> }
    ///
    /// <string_literal> ::= '"' { <letter> | <digit> | <special_char> } '"'
    ///
    /// <literal> ::= <numeral_expr> | <string_literal>
    template <typename T>
    auto parse_literal() -> ast_node;

    /// Parses any expression
    ///
    /// <expr> ::= { <external_linkage> | <function_def> | <primary_expr> }
    auto parse_expr() -> ast_node;

    /// Parses a primary expression
    ///
    /// <primary_expr> ::= <identifier_expr> | <numeral_expr> | <parenthetical_expr> | <binary_expr>
    auto parse_primary_expr() -> ast_node;

    /// Parses a simple identifier expression
    ///
    /// <letter> ::= [A-Za-z]
    /// <digit> ::= [0-9]
    /// <identifier_expr> ::= <letter> { <letter> | <digit> | '_' }
    auto parse_identifier_expr() -> ast_node;

    /// Parses a binary expression, i.e. an expression that evaluates two expressions using
    /// a binary operator.
    ///
    ///  TODO(andrew): support more binary operators
    /// <binary_op> ::= { '*' | '+' | '-' | '/' | '>' | '<' }
    /// <binary_expr> ::= <primary_expr> { <binary_op> <primary_expr> }
    auto parse_binary_expr(std::uint8_t op_precedence = 0) -> ast_node;
    auto parse_binary_expr(ast_node lhs, std::uint8_t op_precedence = 0)
      -> ast_node;

    /// Parses a parenthetical expression, i.e. an expression group with highest evaluation
    /// precedence.
    ///
    /// <parenthetical_expr> ::= '(' <binary_expr> ')'
    auto parse_parenthetical_expr() -> ast_node;

    /// Parses a function argument
    ///
    /// TODO(andrew): add type hints
    /// <function_arg> ::= <identifier>
    auto parse_function_arg() -> function_arg;

    /// Parses a function argument sequence
    ///
    /// <function_arg_seq_multi> ::= <function_arg> | ( <function_arg> ',' <function_arg_seq_multi> )
    /// <function_arg_seq> ::= <empty> | <function_arg_seq_multi>
    auto parse_function_arg_seq() -> std::vector<function_arg>;

    /// Parses a function prototype
    ///
    /// <function_arg> ::= <identifier_seq>
    /// <function_proto> ::= 'function' <identifier_expr> '(' <function_arg_seq> ')'
    auto parse_function_proto() -> function_proto;

    /// Parses a function definition
    ///
    /// <function_def> ::= <function_proto> <primary_expr>
    auto parse_function_def() -> ast_node;

    /// Parses an expression sequence
    ///
    /// <expr_seq_multi> ::= <primary_expr> | ( <primary_expr> ',' <expr_seq_multi> )
    /// <expr_seq> ::= <empty> | <expr_seq_multi>
    auto parse_expr_seq() -> std::vector<ast_node>;

    /// Parses a function call
    ///
    /// <function_call> ::= <identifier_expr> '(' <expr_seq> ')'
    auto parse_function_call(std::string callee) -> ast_node;

    /// Parses an external linkage declaration
    ///
    /// <external_linkage_kind> ::= 'C'
    /// <external_linkage> ::= 'external' '(' <external_linkage_kind> ')' <function_proto>
    auto parse_external_linkage() -> ast_node;

    /// Parses an if/then/else expression
    ///
    /// <else_expr> ::= 'else' <primary_expr>
    /// <if_expr> ::= 'if' <primary_expr> 'then' { <expr> } [ <else_expr> ]
    auto parse_if_expr() -> ast_node;

    /// Parses a type alias
    ///
    /// <type_alias> ::= 'alias' <identifier_expr> 'as' <identifier_expr>
    auto parse_type_alias() -> ast_node;
  };

} // namespace dao
