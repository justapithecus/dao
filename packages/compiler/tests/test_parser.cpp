#include "test_utils.hpp"

#include "parser/parser.hpp"
#include "parser/state_machine.h"

auto directory{Approvals::useApprovalsSubdirectory("golden-files")};

auto main() -> int {

  "identifier_expression_simple"_test = [] {
    auto tokens{load_tokens("identifier_expression_simple.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "numeral_expression"_test = [] {
    auto tokens{load_tokens("numeral_expression.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_simple"_test = [] {
    auto tokens{load_tokens("binary-expressions/simple.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_operator_precedence_1"_test = [] {
    auto tokens{load_tokens("binary-expressions/operator_precedence_1.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "parenthetical_expression"_test = [] {
    auto tokens{load_tokens("parenthetical_expression.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "binary_expression_operator_precedence_2"_test = [] {
    auto tokens{load_tokens("binary-expressions/operator_precedence_2.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_prototype"_test = [] {
    auto tokens{load_tokens("functions/prototype.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_definition"_test = [] {
    auto tokens{load_tokens("functions/definition.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };

  "function_call"_test = [] {
    auto tokens{load_tokens("functions/call.json")};
    auto ast{dao::parse(tokens)};

    json_writer writer{std::move(ast)};
    Approvals::verify(writer);
  };
}
