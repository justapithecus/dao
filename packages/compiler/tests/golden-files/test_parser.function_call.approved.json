{
  "root": {
    "type": "function_call",
    "value": {
      "args": [
        {
          "type": "identifier_expression",
          "value": "p"
        },
        {
          "type": "identifier_expression",
          "value": "q"
        }
      ],
      "callee": "distance_euclidean"
    }
  }
}
